### PR TITLE
21.1 multi-region

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes https://github.com/cockroachlabs/movr-flask/issues/25.
Fixes https://github.com/cockroachlabs/movr-flask/issues/4.

### CODE REVIEW

**Comment by ericharmeling** [comment]
@otan, @rmloveland - added y'all as reviewers, but no rush/feel free to pass to someone else.

I think the main thing I need reviewed is `dbinit.sql`. But feel free to comment on the other changes as well.

File: dbinit.sql [hunk 1]
```diff
@@ -4,130 +4,97 @@ SET sql_safe_updates = false;
...
-ALTER PARTITION europe_west OF INDEX movr.public.rides@rides_auto_index_fk_vehicle_city_ref_vehicles CONFIGURE ZONE USING
-    constraints = '[+region=gcp-europe-west1]';
-ALTER PARTITION us_east OF INDEX movr.public.rides@rides_auto_index_fk_vehicle_city_ref_vehicles CONFIGURE ZONE USING
-    constraints = '[+region=gcp-us-east1]';
-ALTER PARTITION us_west OF INDEX movr.public.rides@rides_auto_index_fk_vehicle_city_ref_vehicles CONFIGURE ZONE USING
-    constraints = '[+region=gcp-us-west1]'
-;
+    CONSTRAINT fk_city_ref_users FOREIGN KEY (rider_id) REFERENCES users(id),
+    CONSTRAINT fk_vehicle_ref_vehicles FOREIGN KEY (vehicle_id) REFERENCES vehicles(id)
+    );
```
Comment (otan) [inline dbinit.sql:82]: nit: unindent ending `);`

File: dbinit.sql [hunk 2]
```diff
@@ -4,130 +4,97 @@ SET sql_safe_updates = false;
...
+
+
+ALTER TABLE rides ADD COLUMN region crdb_internal_region AS (
```
Comment (otan) [inline dbinit.sql:85]: do we ever make this table `REGIONAL BY ROW`? seems like it is REGIONAL BY TABLE at the moment.
we should do this inline, i.e.

```
CREATE TABLE users (

   ...,
   region crdb_internal_region  CASE WHEN city = 'amsterdam' THEN 'gcp-europe-west1'
       WHEN city = 'paris' THEN 'gcp-europe-west1'
       WHEN city = 'rome' THEN 'gcp-europe-west1'
       WHEN city = 'new york' THEN 'gcp-us-east1'
       WHEN city = 'boston' THEN 'gcp-us-east1'
       WHEN city = 'washington dc' THEN 'gcp-us-east1'
       WHEN city = 'san francisco' THEN 'gcp-us-west1'
       WHEN city = 'seattle' THEN 'gcp-us-west1'
       WHEN city = 'los angeles' THEN 'gcp-us-west1'
       ELSE 'gcp-us-east1',
     
       ....
) LOCALITY REGIONAL BY ROW; -- optionally REGIONAL BY ROW AS "region"

same with the others

File: dbinit.sql [hunk 3]
```diff
@@ -4,130 +4,97 @@ SET sql_safe_updates = false;
 DROP DATABASE IF EXISTS movr CASCADE; 
 
 
-CREATE DATABASE movr;
+CREATE DATABASE movr PRIMARY REGION "gcp-us-east1" REGIONS "gcp-europe-west1", "gcp-us-west1";
```
Comment (otan) [inline dbinit.sql:7]: the problem with this setup is that users have to ensure they set up the servers correctly, i.e. set `--locality=region=<blah>` in their flask setup. is that what we want / already have?

File: dbinit.sql (hunk 2 above)
Comment (ericharmeling) [reply dbinit.sql]: My mistake. Fixed for `rides`.

Also, moved the `region` column definition to the `CREATE TABLE` statement for all tables.

File: dbinit.sql (hunk 1 above)
Comment (ericharmeling) [reply dbinit.sql]: After fixing the table (see [comment below](https://github.com/cockroachlabs/movr-flask/pull/27#discussion_r621261601)), I formatted the document with [pgformatter](https://github.com/darold/pgFormatter). Should look nicer all over now.

File: dbinit.sql (hunk 3 above)
Comment (ericharmeling) [reply dbinit.sql]: Yeah ... the README and tutorial specifically instruct the user to specify the localities that match these region names. They are based on the GCP cluster names. On CC, the localities for a multi-region GCP cluster should match these names anyways.

Comment (ericharmeling) [reply dbinit.sql]: TFTR @otan !

**Comment by ericharmeling** [comment]
TFTR @otan !

**Review by rmloveland — APPROVED** [review]
LGTM

As far as I understand what is happening here, is the summary basically the following?

1. Update to use the new multi-region SQL abstractions instead of those yucky partitioning stmts
2. Use a version of movr underneath that does not have the compound PK on (city, id)

**Comment by ericharmeling** [comment]
TFTR @rmloveland !

> 1. Update to use the new multi-region SQL abstractions instead of those yucky partitioning stmts
> 2. Use a version of movr underneath that does not have the compound PK on (city, id)

Exactly. I also snuck in the fixed `cockroach demo` port in the .env and README.

**Pull request merged**